### ISSUE TITLE
[MNG-5440] default-value on mojo parameter of type collection or array effectively make parameter read-only

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6491]
**[Jan Sievers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jan.sievers)** opened **[MNG-5440](https://issues.apache.org/jira/browse/MNG-5440?redirect=false)** and commented

If you use default-value on collection or array mojo parameters like

```java
/**
 * @parameter default-value="foo,bar"
 */
private List<String> param1;
```

the default value will always win over pom.xml configuration.
See demo project attached.
Steps to reproduce:

1. `mvn install` on the demo project
2. `mvn compile -f test-pom.xml -X -e` on the demo project

you can see in the debug log that mojo configuration

```
[DEBUG] Configuration: <?xml version="1.0" encoding="UTF-8"?>
<configuration>
  <param1 default-value="foo,bar">
    <param1>non-default value1 param1</param1>
    <param1>non-default value2 param1</param1>
  </param1>
  <param2 default-value="baz">
    <param2>non-default value1 param2</param2>
    <param2>non-default value2 param2</param2>
  </param2>
</configuration>
```

will not be aplied correctly:

```
[DEBUG] Configuring mojo 'mojoParamsBug:mojoParamsBug:0.0.1-SNAPSHOT:demo' with basic configurator -->
[DEBUG]   (f) param1 = [foo, bar]
[DEBUG]   (f) param2 = [baz]
[DEBUG] -- end configuration --
[INFO] param1: [foo, bar]
[INFO] param2: [baz]
```

If default values for srrays/collection are not supported, this should fail the build in step 1. above.
If they are supported, the configurator should be fixed to actually allow configuring non-default values.

---

**Attachments:**
- [mojoParamsBug.zip](https://issues.apache.org/jira/secure/attachment/12713323/mojoParamsBug.zip) (_11.09 kB_)

**Issue Links:**
- [MPLUGIN-271](https://issues.apache.org/jira/browse/MPLUGIN-271) Usage of defaultValues for a List<..> Parameter seemed not to be working

- [MNG-5001](https://issues.apache.org/jira/browse/MNG-5001) `@readonly` Mojo parameter annotation doesn't work any more

- [MPLUGINTESTING-53](https://issues.apache.org/jira/browse/MPLUGINTESTING-53) default-value on mojo parameter of type collection or array effectively make parameter read-only

- [MPMD-212](https://issues.apache.org/jira/browse/MPMD-212) Add documentation and IT for MPMD-206
 (_**"is depended upon by"**_)

0 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Jan Sievers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jan.sievers)** commented

I'm using maven 3.0.4, same bug happens with SNAPSHOT version of maven 3.1.0

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl@apache.org)** commented

A [fix for this](https://bugs.eclipse.org/bugs/show_bug.cgi?id=470781) has been committed to Sisu Plexus and is scheduled for the 1.0.0 release of that component. For more details see MPLUGIN-271.

Upgrading to Sisu Plexus 1.0.0 should solve this issue.

**Comment by jira-importer** [comment]
**[Stuart McCulloch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mcculls)** commented

I've cherry-picked the fix back to Sisu/Plexus 0.3.2 which is now available on central.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

org.eclipse.sisu.inject-0.3.2 and org.eclipse.sisu.plexus-0.3.2 are part of Maven 3.3.9 distribution, so should be fixed there.
